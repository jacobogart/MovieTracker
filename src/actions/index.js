export const addDiscoverMovies = (movies) => ({
  type: 'ADD_DISCOVER_MOVIES',
  movies
});

export const updateUser = (id, name, favorites) => ({
  type: 'UPDATE_USER',
  id, 
  name,
  favorites
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});
