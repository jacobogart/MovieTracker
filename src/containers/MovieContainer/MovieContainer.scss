.MovieContainer {
  max-width: 1200px;
  min-height: 500px;
  margin: 0 auto;

  .movie-container-header {
    color: white;
    text-align: left;
    margin: 40px 0 10px 25px;
  }

  .display-msg {
    color: rgba(white, .2);
    font-size: 1.6em;
    margin: 50px 0;
  }

  .flex-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
  }
}