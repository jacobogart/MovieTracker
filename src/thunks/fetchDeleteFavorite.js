import { serverUrl } from "../api/pathNames";
import { deleteUserFavorite } from "../actions";

export const fetchDeleteFavorite = (user_id, movie_id) => {
  const url = `${serverUrl}/${user_id}/favorites/${movie_id}`;
  const body = { user_id, movie_id };
  const init = {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  };

  return (dispatch) => {
    fetch(url, init)
      .then(response => {
        if (!response.ok) {
          throw Error("Failed to delete favorite");
        } else {
          return response.json();
        }
      })
      .then(result => {
        console.log(result);
        dispatch(deleteUserFavorite(movie_id));
      })
      .catch(error => console.log(error));
    ;
  };
};