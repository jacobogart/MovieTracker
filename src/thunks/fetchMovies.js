import { jacobKey } from '../utils/API-logins';
import { baseUrl } from '../api/pathNames.js';
import { addDiscoverMovies } from '../actions';
import { cleanAllMovies } from '../api/cleaners';

export const fetchMovies = (fetchCase, query) => {
  let url;
  
  switch(fetchCase) {
    case 'search':
      url = `${baseUrl}search/movie${jacobKey}&language=en-US&query=${query}&page=1&include_adult=false`
      break;
    case 'discover':
      url = `${baseUrl}discover/movie${jacobKey}`
      break;
    default:
      console.log('Need fetchCase');
  } 

  return async (dispatch) => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error("Search failed");
      }
      const data = await response.json();
      const movies = cleanAllMovies(data.results);

      dispatch(addDiscoverMovies(movies));
    } catch(error) {
      console.log(error);
    }
  }
}