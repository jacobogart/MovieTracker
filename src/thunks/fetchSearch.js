import { jacobKey } from '../utils/API-logins';
import { baseUrl } from '../api/pathNames.js';
import { addDiscoverMovies } from '../actions';
import { cleanAllMovies } from '../api/cleaners';

export const fetchSearch =  (query) => {
  const url = `${baseUrl}search/movie${jacobKey}&language=en-US&query=${query}&page=1&include_adult=false`
  return async (dispatch) => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error("Search failed");
      }
      const data = await response.json();
      const movies = cleanAllMovies(data.results)
      dispatch(addDiscoverMovies(movies));
    } catch(error) {
      console.log(error);
    }
  }
}