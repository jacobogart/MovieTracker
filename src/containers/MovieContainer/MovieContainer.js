import React from 'react';
import PropTypes from 'prop-types'
import MovieCard from '../../components/MovieCard/MovieCard';
import { connect } from 'react-redux';
import { updateUserFavorites, deleteUserFavorite } from '../../actions';
import { fetchDeleteFavorite } from '../../thunks/fetchDeleteFavorite';

export const MovieContainer = (props) => {
  let header;
  let message;
  const movieCards = props.movies.map(movie => 
    <MovieCard 
      {...movie} 
      user={props.user} 
      updateUserFavorites={props.updateUserFavorites} 
      deleteUserFavorite={props.deleteUserFavorite}
      key={movie.id} 
    />);
    
  if (props.location.pathname === '/favorites') {

    header = 'Your Favorite Movies'
  } else if (props.location.pathname === '/watchlist') {
    header = 'Your Watchlist'
  } else {
    header = 'Explore Popular Movies'
  }

  if (props.location.pathname === '/favorites' && !props.movies.length) {
    message = <h3 className="display-msg">You have not favorited any movies!</h3>
  } else if (props.location.pathname === '/watchlist' && !props.movies.length) {
    message = <h3 className="display-msg">You have not started a watchlist!</h3>
  }

  return (
    <div className="MovieContainer">
      <h2 className="movie-container-header">{header}</h2>
      <section className="flex-container">
        {message}
        {movieCards}
      </section>
    </div>
  );
}

export const mapStateToProps = (state) => ({
  movies: state.movies,
  user: state.user
});

export const mapDispatchToProps = (dispatch) => ({
  updateUserFavorites: (movieId) => dispatch(updateUserFavorites(movieId)),
  deleteUserFavorite: (userId, movieId) => dispatch(fetchDeleteFavorite(userId, movieId))
})

export default connect(mapStateToProps, mapDispatchToProps)(MovieContainer);

MovieContainer.propTypes = {
  deleteUserFavorite: PropTypes.func,
  history: PropTypes.object,
  location: PropTypes.object,
  match: PropTypes.object,
  movies: PropTypes.array,
  updateUserFavorites: PropTypes.func,
  user: PropTypes.object
};
