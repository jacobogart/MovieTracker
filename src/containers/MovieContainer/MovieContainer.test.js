import React from 'react';
import { shallow } from 'enzyme';
import { cleanMovie, mockUser } from '../../api/__tests__/mockData'
import { MovieContainer, mapStateToProps, mapDispatchToProps } from './MovieContainer';
import { updateUserFavorites, deleteUserFavorite } from '../../actions';
import { fetchDeleteFavorite } from '../../thunks/fetchDeleteFavorite';

const mockMovies = [cleanMovie, cleanMovie];
jest.mock("../../thunks/fetchDeleteFavorite");

describe('MovieContainer', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<MovieContainer movies={mockMovies} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("mapStateToProps", () => {
    it("should return an object with a movies and todos array", () => {
      const state = { user: mockUser, movies: mockMovies, test: 'test' };
      const expected = { user: state.user, movies: state.movies };
      const result = mapStateToProps(state);
      expect(result).toEqual(expected);
    });
  });

  describe("mapDispatchToProps", () => {
    it("should call dispatch when updateUserFavorites is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = updateUserFavorites(11111);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateUserFavorites(11111);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("should call dispatch when deleteUserFavorite is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchDeleteFavorite(1, 11111);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.deleteUserFavorite(1, 11111);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});