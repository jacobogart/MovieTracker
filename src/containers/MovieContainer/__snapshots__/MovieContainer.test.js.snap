// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContainer should match snapshot 1`] = `
<div
  className="MovieContainer"
>
  <h2
    className="movie-container-header"
  >
    Your Favorite Movies
  </h2>
  <section
    className="flex-container"
  >
    <MovieCard
      backdropImg="https://image.tmdb.org/t/p/w500/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg"
      id={299534}
      key="299534"
      overview="After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in order to undo Thanos' actions and restore order to the universe once and for all, no matter what consequences may be in store."
      posterImg="https://image.tmdb.org/t/p/w500/or06FN3Dka5tukK1e9sl16pB3iy.jpg"
      rating={8.6}
      releaseDate="2019-04-24"
      title="Avengers: Endgame"
    />
    <MovieCard
      backdropImg="https://image.tmdb.org/t/p/w500/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg"
      id={299534}
      key="299534"
      overview="After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in order to undo Thanos' actions and restore order to the universe once and for all, no matter what consequences may be in store."
      posterImg="https://image.tmdb.org/t/p/w500/or06FN3Dka5tukK1e9sl16pB3iy.jpg"
      rating={8.6}
      releaseDate="2019-04-24"
      title="Avengers: Endgame"
    />
  </section>
</div>
`;
